aequilibrae/paths/traffic_assignment.py

```diff
@@ -306,3 +329,126 @@ def __validate_parameters(self, kwargs) -> bool:
+                  'VOC_AB', 'VOC_BA', 'VOC_max',
+                  'PCE_AB', 'PCE_BA', 'PCE_tot']
+
+        types = [np.float64] * len(fields)
+        agg = AequilibraeData()
+        agg.create_empty(memory_mode=True, entries=entries, field_names=fields, data_types=types)
+        agg.data.fill(np.nan)
+        agg.index[:] = res1.data.index[:]
+
+        link_ids = class1.results.lids
+        ABs = class1.results.direcs > 0
+        BAs = class1.results.direcs < 0
+
+        indexing = np.zeros(int(link_ids.max()) + 1, np.uint64)
+        indexing[agg.index[:]] = np.arange(entries)
+
+        # Indices of links BA and AB
+        ab_ids = indexing[link_ids[ABs]]
+        ba_ids = indexing[link_ids[BAs]]
+
+        agg.data['Congested_Time_AB'][ab_ids] = np.nan_to_num(self.congested_time[ABs])
+        agg.data['Congested_Time_BA'][ba_ids] = np.nan_to_num(self.congested_time[BAs])
+        agg.data['Congested_Time_Max'][:] = np.nanmax([agg.data.Congested_Time_AB, agg.data.Congested_Time_BA], axis=0)
+
+        agg.data['Delay_factor_AB'][ab_ids] = np.nan_to_num(self.congested_time[ABs] / self.free_flow_tt[ABs])
+        agg.data['Delay_factor_BA'][ba_ids] = np.nan_to_num(self.congested_time[BAs] / self.free_flow_tt[BAs])
+        agg.data['Delay_factor_Max'][:] = np.nanmax([agg.data.Delay_factor_AB, agg.data.Delay_factor_BA], axis=0)
+
+        agg.data['VOC_AB'][ab_ids] = np.nan_to_num(voc[ABs])
+        agg.data['VOC_BA'][ba_ids] = np.nan_to_num(voc[BAs])
+        agg.data['VOC_max'][:] = np.nanmax([agg.data.VOC_AB, agg.data.VOC_BA], axis=0)
+
+        agg.data['PCE_AB'][ab_ids] = np.nan_to_num(tot_flow[ABs])
+        agg.data['PCE_BA'][ba_ids] = np.nan_to_num(tot_flow[BAs])
+        agg.data['PCE_tot'][:] = np.nansum([agg.data.PCE_AB, agg.data.PCE_BA], axis=0)
+
+        assig_results.append(agg)
+
+        dfs = [pd.DataFrame(aed.data) for aed in assig_results]
+        dfs = [df.rename(columns={'index': 'link_id'}).set_index('link_id') for df in dfs]
+        df = pd.concat(dfs, axis=1)
+
+        return df
+
+    def info(self) -> dict:
+        """ Returns information for the traffic assignment procedure
+
+        Dictionary contains keys  'Algorithm', 'Classes', 'Computer name', 'Procedure ID',
+        'Maximum iterations' and 'Target RGap'.
+
+        The classes key is also a dictionary with all the user classes per traffic class and their respective
+        matrix totals
+
+        Returns:
+            *info* (:obj:`dict`): Pandas dataframe with all the assignment results indexed on link_id
+        """
+
+        classes = {}
+        for cls in self.classes:
+            if len(cls.matrix.view_names) == 1:
+                classes[cls.graph.mode] = {nm: np.sum(cls.matrix.matrix_view[:, :]) for nm in cls.matrix.view_names}
+            else:
+                classes[cls.graph.mode] = {nm: np.sum(cls.matrix.matrix_view[:, :]) for i, nm in
+                                           enumerate(cls.matrix.view_names)}
```

Comment:
```suggestion
        def f(cls):
            return {nm: np.sum(cls.matrix.matrix_view[:, :]) for i,nm in enumerate(cls.matrix.view_names)}
        classes = {cls.graph.mode: f(cls) for cls in self.classes}            
```

Reply:
I was just refactoring to show off how I've been using nested functions and dict comprehensions.... but then I couldn't figure out why you even have an if statement or why you are using enumerate

Reply:
Because of that.  The array is two dimensional if only one user class per traffic class, and three dimensional when there are multiple user classes per traffic class

Reply:
Are you forgetting to sum over the third dimension?

Reply:
as in 
```
classes[cls.graph.mode] = { nm: np.sum(cls.matrix.matrix_view[:, :, i]) 
                         for i, nm in enumerate(cls.matrix.view_names)
                       }

Reply:
each dimension is a different slice that gets assigned within a single class, so we can assign multiple demand slices per traffic class

Reply:
Your suggestion doesn't work because of this variation in the number of dimensions of the array, @jamiecook

Reply:
suggestion (anon method and dict comprehension) would still work... just needs your bug fix and the if statement

Reply:
... also... why not just always have a 3 dim matrix?
i.e. `matrix = np.ndarray(x,x,1)`

Reply:
I don't remember now...  There was a reason when I implemented it...